In [2]:
import pandas as pd

# Load dataset spam
spam_df = pd.read_csv('spam.csv', encoding='latin-1')

# Encode label
spam_df['Labels'] = spam_df['v1'].map({
    'ham': 0,
    'spam': 1
})

# Cek data
print(spam_df.head())
print('\nJumlah data:')
print(spam_df.shape)

print('\nDistribusi label:')
print(spam_df['Labels'].value_counts())

print('\nJumlah missing value pada Labels:')
print(spam_df['Labels'].isna().sum())

     v1                                                 v2 Unnamed: 2  \
0   ham  Go until jurong point, crazy.. Available only ...        NaN   
1   ham                      Ok lar... Joking wif u oni...        NaN   
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...        NaN   
3   ham  U dun say so early hor... U c already then say...        NaN   
4   ham  Nah I don't think he goes to usf, he lives aro...        NaN   

  Unnamed: 3 Unnamed: 4  Labels  
0        NaN        NaN       0  
1        NaN        NaN       0  
2        NaN        NaN       1  
3        NaN        NaN       0  
4        NaN        NaN       0  

Jumlah data:
(5572, 6)

Distribusi label:
Labels
0    4825
1     747
Name: count, dtype: int64

Jumlah missing value pada Labels:
0


In [3]:
from sklearn.model_selection import train_test_split

X = spam_df['v2'].to_numpy()
y = spam_df['Labels'].to_numpy()

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=50
)

print("Jumlah data training:", len(X_train))
print("Jumlah data testing :", len(X_test))

Jumlah data training: 4457
Jumlah data testing : 1115


In [4]:
from sklearn.feature_extraction.text import CountVectorizer

bow = CountVectorizer(stop_words='english')

X_train_bow = bow.fit_transform(X_train)
X_test_bow = bow.transform(X_test)

print("Ukuran X_train:", X_train_bow.shape)
print("Ukuran X_test :", X_test_bow.shape)

Ukuran X_train: (4457, 7466)
Ukuran X_test : (1115, 7466)


In [5]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

mnb_bow = MultinomialNB()

mnb_bow.fit(X_train_bow, y_train)

y_pred_bow = mnb_bow.predict(X_test_bow)

accuracy_bow = accuracy_score(y_test, y_pred_bow)

print("Accuracy:", accuracy_bow)

print("\nClassification Report:")
print(classification_report(y_test, y_pred_bow))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_bow))

Accuracy: 0.9829596412556054

Classification Report:
              precision    recall  f1-score   support

           0       0.98      1.00      0.99       954
           1       0.98      0.90      0.94       161

    accuracy                           0.98      1115
   macro avg       0.98      0.95      0.96      1115
weighted avg       0.98      0.98      0.98      1115


Confusion Matrix:
[[951   3]
 [ 16 145]]


In [6]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf = TfidfVectorizer(stop_words='english')

X_train_tfidf = tfidf.fit_transform(X_train)
X_test_tfidf = tfidf.transform(X_test)

print("Ukuran X_train TF-IDF:", X_train_tfidf.shape)
print("Ukuran X_test TF-IDF :", X_test_tfidf.shape)

Ukuran X_train TF-IDF: (4457, 7466)
Ukuran X_test TF-IDF : (1115, 7466)


In [7]:
mnb_tfidf = MultinomialNB()

mnb_tfidf.fit(X_train_tfidf, y_train)

y_pred_tfidf = mnb_tfidf.predict(X_test_tfidf)

accuracy_tfidf = accuracy_score(y_test, y_pred_tfidf)

print("Accuracy:", accuracy_tfidf)

print("\nClassification Report:")
print(classification_report(y_test, y_pred_tfidf))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_tfidf))

Accuracy: 0.9605381165919282

Classification Report:
              precision    recall  f1-score   support

           0       0.96      1.00      0.98       954
           1       1.00      0.73      0.84       161

    accuracy                           0.96      1115
   macro avg       0.98      0.86      0.91      1115
weighted avg       0.96      0.96      0.96      1115


Confusion Matrix:
[[954   0]
 [ 44 117]]


# 1. Construct Multinomial Naive Bayes using CountVectorizer with stop words

A Multinomial Naive Bayes classification model was constructed to classify SMS messages into two classes: ham and spam.

The dataset was divided into 80% training data and 20% testing data, resulting in 4,457 training samples and 1,115 testing samples.

For feature extraction, CountVectorizer with English stop words removal was used. The resulting feature matrix contained 7,466 features.

The Multinomial Naive Bayes model achieved an accuracy of:

98.30%


The model correctly classified 951 ham messages and 145 spam messages. There were 3 ham messages incorrectly classified as spam and 16 spam messages incorrectly classified as ham.

The classification report showed an F1-score of 0.94 for the spam class, indicating that the model performed very well in detecting spam messages.



# 2. Construct Multinomial Naive Bayes using TF-IDF with stop words

In the second experiment, TF-IDF Vectorizer with English stop words removal was used as the feature extraction method.

The training and testing data remained the same as in the first experiment. The resulting TF-IDF feature matrix also contained 7,466 features.

The Multinomial Naive Bayes model achieved an accuracy of:

96.05%



The model correctly classified all 954 ham messages, but it incorrectly classified 44 spam messages as ham. Therefore, the recall of the spam class decreased to 73%, with an F1-score of 84%.


# 3. Comparison and conclusion


Based on the experimental results, CountVectorizer with stop words removal is the optimal feature extraction method for this dataset.

CountVectorizer achieved a higher accuracy of 98.30%, compared with 96.05% for TF-IDF. It also achieved better spam recall and F1-score. In particular, CountVectorizer correctly detected 145 out of 161 spam messages, while TF-IDF correctly detected only 117 out of 161 spam messages.

Therefore, for the `spam.csv` dataset and the Multinomial Naive Bayes model used in this experiment, CountVectorizer with stop words removal provides better classification performance than TF-IDF with stop words removal.